### DISCUSSION TITLE
Feature Request - Booleans

### DISCUSSION THREAD

**Discussion by YarlBoro** [guillaumechereau/goxel#351] [Ideas]
As the title says, I would like to propose implementing boolean operations between contents of two layers. **Intersection** - remove volume shared by two layers or keep only volume shared by two selected layers. **Substraction** - remove voxels of one layer intersecting with another layer (cut one with another one). Hell, maybe even real time booleans with current parametric shape layers would be possible? Booleans are a very powerful feature substantialy speeding up the modeling process. Things that normaly take many, many steps can be applied in a matter of seconds. Esspecially when preparing man made objects like vehicles or weapons. I can help with UX design & graphics if that would be of any use. BTW, new UI is f*****g awesome :)

**Comment by guillaumechereau** [comment]
Yes that would be nice.  Maybe adding a layer 'mode' combo box similar to what gimp has.  It could be boolean operation, or even maybe color transfer functions?

**Comment by YarlBoro** [comment]
Sounds terrific! Color transfer functions - I had to check what this implies and that looks as (damn) useful as booleans.

**Comment by guillaumechereau** [comment]
I am not sure there is actually a need for that… I guess I’ll start with a
test if adding a “substract” and “intersect” mode, see if that works.

**Comment by YarlBoro** [comment]
If I may suggest one more thing, please consider exposing functionalities for inter-layers operations for scripting. IMHO it would open up lots of venues.
If you need any help with testing or enything else, let me know.

**Comment by guillaumechereau** [comment]
Scripting is a bit tricky.  I added some basic support for javascript scripting, but now I realize lua might have been a better idea (still unsure about it).  If you can make some moc-up of what you have in mind I can have a look.

**Comment by YarlBoro** [comment]
I see, its tricky indeed.
Sure, I will take care of it.

**Comment by guillaumechereau** [comment]
I just pushed a first commit for this feature.  You can test it out with the automatic build from github:
https://github.com/guillaumechereau/goxel/actions/runs/7801466760

![image](https://github.com/guillaumechereau/goxel/assets/107679/2f760e9e-a604-410b-93c3-4e2a1ae46e9f)

**Comment by YarlBoro** [comment]
It is undestructive, wow. I imagined sth more basic. No complains here, this is perfect!

In quick tests all works good. I will see if it breaks on more complicated models once I'm home.

**Comment by YarlBoro** [comment]
What do you think about this?
![booleans_v01_01](https://github.com/guillaumechereau/goxel/assets/7037937/a6953ffb-f7e2-419f-b22f-1245a307e3bc)

**Comment by YarlBoro** [comment]
I've played first with 8-10 smaller objects, after that with 5-6 big chunky objects in the work space of 128x128x128 in various configurations and boolean modes. I wasn't able to break it Goxel, I've really tried :) There are slow downs when I moved any element that was part of the booleans chain, but I think this is to be expected. While using booleans in the project I had no problem with other tools so far. No glitches, freezes or hangs. If you have suggestions in regard to the methodology, let me know.

I'm not sure if order of boolean actions should go from the bottom up (layers) as it does now. Up to bottom is well established convention in 2D/3D editors. ATM it is a bit confusing.

I know this is just an early test, I wasn't expecting a function to merge booleans into the static voxels. I wonder how it would look like function & UI wise. What should be merged? All of the visible layers? Everything above/under selected layer? I think it would be simpler if there would be folders/groups of layers in the Goxel. User could put set of layers with boolean operations between them into the separate folder and when the time comes uses function to merge the whole group/folder. But folders for layers are a feature on its own, probably requiring lots of effort.

Let me know what you think about draft of the UI.

**Comment by guillaumechereau** [comment]
Yeah unfortunately at the moment layers group would be a bit complicated.
What I’ll add is a button to « merge down » a layer since I think that’s
the flow that would be the simpler: create a layer with Boolean operation
and once satisfied, merge it to avoid having a long operation chain.

Should probably show something in the layer list to to clearly identify the
layers with special modes.

**Reply by YarlBoro** [reply]
Gocha, makes sense.

Identify layers with special modes - I think icon symbolizing it could appear on the rightmost side of the layers name field. I will come up with a proposition of an icon and UI mockup.

**Comment by YarlBoro** [comment]
Simple mockup with two versions of icon marking that layer has a special mode active. Upper one is generalized boolean icon similar to  for buttons, bottom one suppose to suggest a logic gate. 

![layer_booleans_icon_mockup_001](https://github.com/guillaumechereau/goxel/assets/7037937/16051f43-64c3-41fc-b58f-3dedb332e02e)
